src/lib/syncable_ledger/syncable_ledger.ml

```diff
@@ -427,58 +532,90 @@ end = struct
     if Hash.equal actual expected then `Success
     else `Hash_mismatch (expected, actual)
 
-  (** Given an address and the hashes of the children of the corresponding node,
-      check the children hash to the expected value. If they do, queue the
-      children for retrieval if the values in the underlying ledger don't match
-      the hashes we got from the network. *)
-  let add_child_hashes_to :
+  (* Merges each 2 contigous nodes, halving the size of the array *)
+  let merge_siblings : Hash.t array -> index -> Hash.t array =
+   fun nodes height ->
+    let len = Array.length nodes in
+    if len mod 2 <> 0 then failwith "length must be even" ;
+    let half_len = len / 2 in
+    let f i = Hash.merge ~height nodes.(2 * i) nodes.((2 * i) + 1) in
+    Array.init half_len ~f
+
+  (* Assumes nodes to be a power of 2 and merges them into their common root *)
+  let rec merge_many : Hash.t array -> index -> Hash.t =
+   fun nodes height ->
+    let len = Array.length nodes in
+    match len with
+    | 1 ->
+        nodes.(0)
+    | _ ->
+        let half = merge_siblings nodes height in
+        merge_many half (height + 1)
+
+  let merge_many : Hash.t array -> index -> index -> Hash.t =
+   fun nodes height subtree_depth ->
+    let bottom_height = height - subtree_depth in
+    let hash = merge_many nodes bottom_height in
+    hash
+
+  (* Adds the subtree given as the 2^k subtree leaves with the given prefix address *)
+  (* Returns next nodes to be checked *)
+  let add_subtree :
          'a t
       -> Addr.t
-      -> Hash.t
-      -> Hash.t
-      -> [ `Good of (Addr.t * Hash.t) list
-           (** The addresses and expected hashes of the now-retrievable children *)
+      -> Hash.t array
+      -> int
+      -> [ `Good of (Addr.t * Hash.t) array
          | `Hash_mismatch of Hash.t * Hash.t
-           (** Hash check failed, peer lied. First parameter expected, second parameter actual. *)
-         ] =
-   fun t parent_addr lh rh ->
-    let ledger_depth = MT.depth t.tree in
-    let la, ra =
-      Option.value_exn ~message:"Tried to fetch a leaf as if it was a node"
-        ( Or_error.ok
-        @@ Or_error.both
-             (Addr.child ~ledger_depth parent_addr Direction.Left)
-             (Addr.child ~ledger_depth parent_addr Direction.Right) )
+         | `Invalid_length ] =
+   fun t addr nodes requested_depth ->
+    let (module Context) = t.context in
+    let len = Array.length nodes in
+    let is_power = Int.is_pow2 len in
+    let is_more_than_two = len >= 2 in
+    let subtree_depth = Int.ceil_log2 len in
+    let less_than_max =
+      len <= Int.pow 2 Context.compile_config.sync_ledger_max_subtree_depth
```

Comment:
We don't care if it's less than our max. The only relevant maximum is the sending node's maximum.

Reply:
I was thinking the receiving node wouldn't want to process some arbitrarily long list of hashes. Otherwise what is the need for specifying depth in the query if receiver accepts any depth?

Reply:
The other test above checks that it's less than what we requested. The max is just what we're locally happy to respond with, unrelated to the responses we receive.

Reply:
Oh wait, I added the requested depth check and this comment is for max. Sorry, you're right, we don't need the max